askup/models.py

```diff
@@ -1,3 +1,172 @@
+    parent_qset = models.ForeignKey(
+        "self",
+        on_delete=models.CASCADE,
+        null=True,
+        blank=True,
+        db_index=True
+    )
+    for_any_authenticated = models.BooleanField(default=False, db_index=True)
+    for_unauthenticated = models.BooleanField(default=False, db_index=True)
+    show_authors = models.BooleanField(default=True)
+    own_questions_only = models.BooleanField(default=False)
+    users = models.ManyToManyField(User, blank=True, db_index=True)
+    questions_count = models.PositiveIntegerField(default=0, db_index=True)
+
+    @transaction.atomic
+    def save(self, *args, **kwargs):
+        previous_parent_id = self.parent_qset_id
+        super().save(*args, **kwargs)
+
+        if (self.parent_qset_id != previous_parent_id
+                and self.parent_qset_id
+                and self.questions_count != 0):
+            previous_parent = Qset.objects.get(id=previous_parent_id)
+            previous_parent.iterate_questions_count(-self.questions_count)
+            previous_parent.save()
+            self.parent_qset.iterate_questions_count(self.questions_count)
+
+        super().save(*args, **kwargs)
+
+    @transaction.atomic
+    def delete(self, *args, **kwargs):
+        self.iterate_questions_count(-self.questions_count)
+        super().delete(*args, **kwargs)
+
+    def iterate_questions_count(self, amount):
+        """
+        Iterate questions_count on a specific amount
+        recursively to the top parent (organization)
+        """
+        self.questions_count += amount
+
+        if self.parent_qset_id:
+            self.parent_qset.iterate_questions_count(amount)
+
+        self.save()
+
+    def validate_unique(self, exclude=None):
+        if self.parent_qset_id is None:
+            is_exists = Qset.objects.exclude(id=self.id).filter(
+                name=self.name,
+                parent_qset_id__isnull=True
+            ).exists()
+
+            if is_exists:
+                raise ValidationError("Organization with such name is already exists")
+
+        super().validate_unique(exclude)
+
+    def __str__(self):
+        return self.name
+
+    class Meta:
+        unique_together = ('parent_qset', 'name')
+
+
+class Organization(Qset):
+    """Describes the Organization model and it's behaviour"""
+    def __str__(self):
+        return self.name
+
+
+class EmailPattern(models.Model):
+    """Contains all the email patterns of the organizations"""
+    organization = models.ForeignKey(Organization, on_delete=models.CASCADE, null=True, db_index=True)
+    text = models.CharField(max_length=50, db_index=True)
+
+
+class Question(models.Model):
+    """Describes all the Qset's and the Organization's models and their behaviours"""
+    BLOOMS_TAGS = (
+        (0, 'remember'),
+        (1, 'understand'),
+        (2, 'apply'),
+        (3, 'analyze'),
+        (4, 'evaluate'),
+        (5, 'create')
+    )
+    text = models.TextField(db_index=True)
```

Comment:
Usual recommendation to start using indexing is the case of the performance issues of hight load of the search operations on the certain field why do we index `answer_text `, `qset `, `created_at ` and `updated_at ` as well as `user ` fields?

Reply:
I've removed db_index from the answer_text, created_at and updated_at.

I've removed the `user` index but it doesn't make any difference because all the foreign constraints must be provided with the index of the corresponding fields and Django is ensuring them anyway.

I left `qset` explicitly indexed though because this field will be used a lot in a "qset questions list" where statements.